src/ir.cpp

```diff
@@ -23420,6 +23452,82 @@ static IrInstruction *ir_analyze_instruction_align_of(IrAnalyze *ira, IrInstruct
     return result;
 }
 
+static IrInstruction *ir_analyze_instruction_expect(IrAnalyze *ira, IrInstructionExpect *instruction) {
+    IrInstruction *value = instruction->value->child;
+    if (type_is_invalid(value->value.type))
+        return ira->codegen->invalid_instruction;
+
+    ZigType *value_type = value->value.type;
+
+    IrInstruction *expect_instr = instruction->expected->child;
+    if (type_is_invalid(expect_instr->value.type))
+        return ira->codegen->invalid_instruction;
+
+    switch (value_type->id) {
+        case ZigTypeIdInvalid:
+            zig_unreachable();
+        case ZigTypeIdMetaType:
+        case ZigTypeIdVoid:
+        case ZigTypeIdUnreachable:
+        case ZigTypeIdFloat:
+        case ZigTypeIdArray:
+        case ZigTypeIdComptimeFloat:
+        case ZigTypeIdComptimeInt:
+        case ZigTypeIdUndefined:
+        case ZigTypeIdNull:
+        case ZigTypeIdErrorSet:
+        case ZigTypeIdArgTuple:
+        case ZigTypeIdOpaque:
+        case ZigTypeIdFnFrame:
+        case ZigTypeIdAnyFrame:
+        case ZigTypeIdVector:
+        case ZigTypeIdStruct:
+        case ZigTypeIdUnion:
+        case ZigTypeIdEnumLiteral:
+            {
+                ErrorMsg *msg = ir_add_error(ira, value,
+                    buf_sprintf("type %s not allowed as value to @expect. Allowed types are bool, enum, int, ...",
+                        buf_ptr(&value_type->name)));
+                if (value_type->id == ZigTypeIdEnumLiteral && expect_instr->value.type->id == ZigTypeIdEnum) {
+                    add_error_note(ira->codegen, msg, value->source_node,
+                        buf_sprintf("did you reverse the arguments?"));
+                }
+                return ira->codegen->invalid_instruction;
+            }
+        case ZigTypeIdBoundFn:
+        case ZigTypeIdFn:
+        case ZigTypeIdPointer:
+        case ZigTypeIdOptional:
+        case ZigTypeIdErrorUnion:
+            zig_panic("TODO. more types for @expect()");
```

Comment:
`@expect` should work on all types. Instead of panicking here, just return `value`.